daceml/onnx/op_implementations/fpga_implementations.py

```diff
@@ -2807,17 +2799,187 @@ def forward(node: ONNXOp, state: SDFGState,
                                   memlet=dace.Memlet("sum_res[0]"))
         new_state.add_memlet_path(outer_me, init_tasklet, memlet=dace.Memlet())
 
-
         new_state.add_memlet_path(store_tasklet,
                                   outer_mx,
                                   out_data,
                                   src_conn="out_res",
                                   memlet=dace.Memlet("reduced[o0, o1, o2]"))
 
-
-
-
         new_sdfg.fill_scope_connectors()
         new_sdfg.validate()
         new_sdfg.save('/tmp/reduce_sum.sdfg')
         return new_sdfg
+
+
+# -----------------------------
+# Binary element-wise operators
+# -----------------------------
+def binary_forward(node: ONNXOp, state: SDFGState, sdfg: SDFG,
+                   operator: str) -> typing.Union[Node, SDFG]:
+
+    node.validate(sdfg, state)
+
+    op_map = {'+': 'add', '-': 'sub', '*': 'mul', '/': 'div'}
+    op_name = op_map[operator]
+
+    streaming_node = False
+
+    # Get inputs and outputs names
+    A = in_desc_with_name(node, state, sdfg, "A")
+    B = in_desc_with_name(node, state, sdfg, "B")
+    C = out_desc_with_name(node, state, sdfg, "C")
+
+    # Create inner SDFG
+    op_sdfg = dace.SDFG(node.name)
+    compute_state = op_sdfg.add_state("compute_{}".format(op_name))
+
+    a_inner = A.clone()
+    a_inner.transient = False
+    b_inner = B.clone()
+    b_inner.transient = False
+    c_inner = C.clone()
+    c_inner.transient = False
+
+    op_sdfg.add_datadesc("A", a_inner)
+    op_sdfg.add_datadesc("B", b_inner)
+    op_sdfg.add_datadesc("C", c_inner)
+
+    # Create map and tasklet
+    map_ranges = {'__i%d' % i: '0:%s' % n for i, n in enumerate(A.shape)}
+    outer_me, outer_mx = compute_state.add_map(
+        '{}_map'.format(op_name),
+        map_ranges,
+        schedule=dace.ScheduleType.FPGA_Device)
+
+    tasklet = compute_state.add_tasklet(
+        '{}_task'.format(op_name), ['a_con', 'b_con'], ['c_con'],
+        'c_con = a_con {} b_con'.format(operator))
+
+    # Add data descriptors
+    a_in = compute_state.add_read("A")
+    b_in = compute_state.add_read("B")
+    c_out = compute_state.add_write("C")
+
+    # Connect map
+    if not streaming_node:
```

Comment:
I would suggest refactoring this, since (as far as I can see) the only thing changing is the subset, to something like:

```python
if not streaming_node:
    subset = ", ".join(f"{i}" for i in range(len(A.shape)))
else:
    subset = "0"

compute_state.add_memlet_path(
    a_in,
    outer_me,
    tasklet,
    dst_conn="a_conn",
    memlet=dace.Memlet(f"A[{subset}]"))
```

etc.